pkg/util/ciutil/systems.go

```diff
@@ -14,35 +14,106 @@
 	// as a fallback so that the CLI would still pick-up the metadata related
 	// to their CI build.
-	GenericCI System = "Generic CI"
-
-	GitHub      System = "GitHub"
-	GitLab      System = "GitLab"
-	GoCD        System = "GoCD"
-	Hudson      System = "Hudson"
-	Jenkins     System = "Jenkins"
-	MagnumCI    System = "Magnum CI"
-	Semaphore   System = "Semaphore"
-	TaskCluster System = "TaskCluster"
-	TeamCity    System = "TeamCity"
-	Travis      System = "Travis CI"
+	GenericCI SystemName = "Generic CI"
+
+	GitHub      SystemName = "GitHub"
+	GitLab      SystemName = "GitLab"
+	GoCD        SystemName = "GoCD"
+	Hudson      SystemName = "Hudson"
+	Jenkins     SystemName = "Jenkins"
+	MagnumCI    SystemName = "Magnum CI"
+	Semaphore   SystemName = "Semaphore"
+	TaskCluster SystemName = "TaskCluster"
+	TeamCity    SystemName = "TeamCity"
+	Travis      SystemName = "Travis CI"
 )
 
-// System is a recognized CI system.
-type System string
+// SystemName is a recognized CI system.
+type SystemName string
+
+// system represents a CI/CD system.
+type system interface {
+	// DetectVars when called on a specific instance of a CISystem
+	// detects the env vars of the corresponding CI/CD system and
+	// returns `Vars` with those values.
+	DetectVars() Vars
+	// IsCI returns true if any of the CI systems's associated environment variables are set.
+	IsCI() bool
+}
+
+// Vars contains a set of metadata variables about a CI system.
+type Vars struct {
+	// Name is a required friendly name of the CI system.
+	Name SystemName
+	// BuildID is an optional unique identifier for the current build/job.
+	BuildID string
+	// BuildType is an optional friendly type name of the build/job type.
+	BuildType string
+	// BuildURL is an optional URL for this build/job's webpage.
+	BuildURL string
+	// SHA is the SHA hash of the code repo at which this build/job is running.
+	SHA string
+	// BranchName is the name of the feature branch currently being built.
+	BranchName string
+	// CommitMessage is the full message of the Git commit being built.
+	CommitMessage string
+	// PRNumber is the pull-request ID/number in the source control system.
+	PRNumber string
+}
+
+// baseCI implements the `System` interface with default
+// implementations.
+//
+// When creating a new CI System implementation, implement the
+// DetectVars and any other function you wish to override.
+type baseCI struct {
+	Name SystemName
+	// EnvVarsToDetect is an array of env vars to check if any of these env vars is set,
+	// which would indicate that the Pulumi CLI is running in that CI system's environment.
+	EnvVarsToDetect []string
+	// EnvValuesToDetect is a map of env vars and their expected values to check for,
+	// in order to see if the Pulumi CLI is running inside a certain CI system's environment.
+	EnvValuesToDetect map[string]string
+}
+
+// DetectVars for a known CI system returns an empty Vars struct.
+func (d baseCI) DetectVars() Vars {
+	return Vars{}
```

Comment:
For the `baseCI` case, we might as return the `Vars { Name: baseCI.Name }` instead of an entirely empty `Vars` struct. Since we at least know the name of the CI/CD system used.

Reply:
Very good catch! This should be `return Vars{Name: d.Name}`

Reply:
I am adding a test case to cover the `baseCI` scenario.